readthedocs/builds/tests/test_views.py

```diff
@@ -0,0 +1,88 @@
+from unittest import mock
+
+from django.contrib.auth.models import User
+from django.test import TestCase
+from django.urls import reverse
+from django_dynamic_fixture import get
+
+from readthedocs.builds.constants import (
+    BUILD_STATE_CANCELLED,
+    BUILD_STATE_INSTALLING,
+    BUILD_STATE_TRIGGERED,
+)
+from readthedocs.builds.models import Build, Version
+from readthedocs.projects.models import Project
+
+
+@mock.patch("readthedocs.builds.views.app")
+class CancelBuildViewTests(TestCase):
+    def setUp(self):
+        self.user = get(User, username="test")
+        self.project = get(Project, users=[self.user])
+        self.version = get(Version, project=self.project)
+        self.build = get(
+            Build,
+            project=self.project,
+            version=self.version,
+            task_id="1234",
+            state=BUILD_STATE_INSTALLING,
+        )
+
+    def test_cancel_running_build(self, app):
+        self.build.state = BUILD_STATE_INSTALLING
+        self.build.save()
+        self.client.force_login(self.user)
+        url = reverse("builds_detail", args=[self.project.slug, self.build.pk])
+        resp = self.client.post(url)
+        self.assertEqual(resp.status_code, 302)
+        app.control.revoke.assert_called_once_with(
+            self.build.task_id, signal=mock.ANY, terminate=True
+        )
+        self.build.refresh_from_db()
+        self.assertEqual(self.build.state, BUILD_STATE_INSTALLING)
```

Comment:
Why is this still `installing` at this point?

Reply:
Changing the state of the build object will be handled inside the task

https://github.com/readthedocs/readthedocs.org/blob/4983fce383ce4b859f1c8636f139a657cfc03c7d/readthedocs/projects/tasks/builds.py#L274-L276